File: temporal/ui/video_renderer_editor.py
from typing import Iterator

import gradio as gr

from temporal.ui import ReadData, UIThing, UpdateData, UpdateRequest, Widget
from temporal.ui.gradio_widget import GradioWidget
from temporal.ui.reorderable_list import ReorderableList
from temporal.ui.video_filter_editor import VideoFilterEditor
from temporal.utils.collection import find_by_predicate, reorder_dict
from temporal.video_renderer import VideoRenderer


class VideoRendererEditor(Widget):
    def __init__(
        self,
        value: VideoRenderer = VideoRenderer(),
    ) -> None:
        super().__init__()

        self._fps = GradioWidget(gr.Slider, label = "Frames per second", minimum = 1, maximum = 60, step = 1, value = value.fps)

        with GradioWidget(gr.Row):
            self._first_frame = GradioWidget(gr.Number, label = "First frame", precision = 0, minimum = 1, step = 1, value = value.first_frame)
            self._last_frame = GradioWidget(gr.Number, label = "Last frame", precision = 0, minimum = 0, step = 1, value = value.last_frame)

        self._frame_stride = GradioWidget(gr.Number, label = "Frame stride", precision = 0, minimum = 1, step = 1, value = value.frame_stride)
        self._looping = GradioWidget(gr.Checkbox, label = "Looping", value = value.looping)

        with ReorderableList() as self._filter_order:
            self._filters = {
                filter.id: VideoFilterEditor(value = filter)
                for filter in value.filters
            }

    @property
    def dependencies(self) -> Iterator[UIThing]:
        yield self._fps
        yield self._first_frame
        yield self._last_frame
        yield self._frame_stride
        yield self._looping
        yield self._filter_order
        yield from self._filters.values()

    def read(self, data: ReadData) -> VideoRenderer:
        return VideoRenderer(
            fps = data[self._fps],
            first_frame = data[self._first_frame],
            last_frame = data[self._last_frame],
            frame_stride = data[self._frame_stride],
            looping = data[self._looping],
            filters = [
                data[widget]
                for widget in reorder_dict(self._filters, [
                    self._filter_ids[x]
                    for x in data[self._filter_order]
                ]).values()
            ],
        )

    def update(self, data: UpdateData) -> UpdateRequest:
        result: UpdateRequest = {}

        if isinstance(value := data.get("value", None), VideoRenderer):
            result[self._fps] = {"value": value.fps}
            result[self._first_frame] = {"value": value.first_frame}
            result[self._last_frame] = {"value": value.last_frame}
            result[self._frame_stride] = {"value": value.frame_stride}
            result[self._looping] = {"value": value.looping}
            result[self._filter_order] = {"value": [self._filter_ids.index(x.id) for x in value.filters]}
            result |= {
                widget: {"value": find_by_predicate(value.filters, lambda x: x.id == id)}
                for id, widget in self._filters.items()
            }

        return result

    @property
    def _filter_ids(self) -> list[str]:
        return list(self._filters.keys())

File: ui/scripts/video_renderer_editor.js
import {Checkbox} from "../scripts/base/checkbox.js";
import {Form} from "../scripts/base/form.js";
import {NumberBox} from "../scripts/base/number_box.js";
import {Slider} from "../scripts/base/slider.js";
import {FieldManager} from "../scripts/core/field_manager.js";
import {Signal} from "../scripts/core/signal.js";
import {mapObject} from "../scripts/utils/object.js";
import {ModuleList} from "../scripts/module_list.js";
import {videoFilters} from "../scripts/shared_data.js";
import {VideoFilterEditor} from "../scripts/video_filter_editor.js";

export class VideoRendererEditor extends Form {
    constructor() {
        super();

        this.onValueChange = new Signal();

        this._manager = new FieldManager(this.onValueChange);

        this.createField("Frames per second", Slider, (e) => {
            e.minimum = 1;
            e.maximum = 60;
            e.step = 1;
            e.value = 30;
            this._manager.manage(e, "fps");
        });

        this.createRow((e) => {
            e.createField("First frame", NumberBox, (e) => {
                e.minimum = 1;
                e.step = 1;
                e.value = 1;
                this._manager.manage(e, "first_frame");
            });

            e.createField("Last frame", NumberBox, (e) => {
                e.minimum = 0;
                e.step = 1;
                e.value = 0;
                this._manager.manage(e, "last_frame");
            });
        });

        this.createField("Frame stride", NumberBox, (e) => {
            e.minimum = 1;
            e.step = 1;
            e.value = 1;
            this._manager.manage(e, "frame_stride");
        });

        this.createField("Looping", Checkbox, (e) => {
            e.value = false;
            this._manager.manage(e, "looping");
        });

        this.createField("Add filter", ModuleList, (e) => {
            this._manager.manage(e, "filters");
        }, VideoFilterEditor, mapObject(videoFilters, (_, filter) => filter.name), videoFilters);

        this.createField("Archive mode", Checkbox, (e) => {
            e.value = false;
            this._manager.manage(e, "archive_mode");
        });
    }

    get value() {
        return this._manager.value;
    }

    set value(value) {
        this._manager.value = value;
    }
}
customElements.define("video-renderer-editor", VideoRendererEditor);
